### DISCUSSION TITLE
Entry of small datasets directly into R-Instat

### DISCUSSION THREAD

**Discussion by rdstern** [IDEMSInternational/R-Instat#7130] [Ideas]
The first dialogue in the first R-Instat menu is File > New Data Frame.  This is therefore where the reviewer (Bob Muenchen) started and he was very critical of this bit of R-Instat.
Now I have worked very hard, with @Patowhiz, on this dialogue and claim it is brilliant (of course)!  So why is Bob so critical and why does it need to change?
One of Bob's points is that it is highly daunting for the majority of his supposed audience for an R GUI.  The large audience is those who are looking for a GUI because they don't want to see R commands - ever!  And here the first dialogue on the first menu is supposed to excite them by showing how just a little use of R-commands is within easy reach of many potential users and that's pretty impressive.  In contrast, they are pretty frightened.  If that is what the start is like, how difficult is it going to be overall.

Our initial response was to agree that the empty data frame should be the default.  That will already make a big difference and will be easy to do.

![image](https://user-images.githubusercontent.com/11226469/149732616-e6a52a18-28cf-425f-8ccd-057f7c395a5e.png)

So, we simply make Empty the first, rather than the last, option.  That's in the dialogue above and is easy to do, and our first task.

But this is a discussion topic - so we discussed with Bob.  Our view is that you will almost never need this option.  There are so many other ways of getting the data into R-Instat.  (You could type it into Excel, or into SPSS.  You could use a proper data entry system, like Epiinfo or CSPRO).   
And another point is that statistics teaching is messed up by too much time being spent on theory and then simplistic (and uninteresting) data sets to illustrate the theory.  So, this option will be used so rarely that we should go further and move the File New down in the menu - so it isn't the first item - and therefore an obvious starting point.  

Bob's reply is that, in his experience, many people do start by entering a small dataset directly into the GUI.  The implications is "What's the point of a GUI if you can't simply enter your data there."  Now we may wish the world is a bit different, but perhaps that is a bit of our problem with R-Instat.  We should at least start with the world as it is.

He takes this further to a possible audience, namely students on "Statistics 101".  So a first course in statistics for a large class, many of whom might do no further courses in statistics.  And they could be pretty negative about the subject.

Their tasks could well include some small data sets, and they may have to be entered.  I have always assumed they would probably enter into a spreadsheet and then use File > Import.  But that assumes users are happy with the concept of more than one package from the start.  Also that they are probably migrating from a spreadsheet.  In some courses, and with some people they won't be - and this introduces an unnecessary complication.  So they should be able to enter directly into R-Instat.  We have given the option.  We should follow-through and "make it easy"!

So here is a small dataset from the Mosaic package.  I tried typing it in and found a few more things we could usefully change.  That's also what I would like to discuss here.

![image](https://user-images.githubusercontent.com/11226469/149737438-75e5daaa-3011-4aef-aa57-46f15fe22b8e.png)
a) So I made the empty dataset using the File > New dialogue.  
b) Then right-click and Rename Column.  I had to do it twice.  ok, 
c) So could we add variable names as an option on the File > New dialogue? It would open a small grid where you could type the variable names.  
d) What else - well perhaps Variable Labels as well.  And Column types - at least Logical (the default), numeric or Character.  
e) It will be too much work to allow factor at this stage.
f) Then I typed some data.  That was sort of ok.
g) But then I made a mistake.  I found we don't have delete yet.  I strongly suggest we permit delete cell and block within the grid- at least within a variable.  It could give a warning.  It would replace the value(s) by NA.  (We don't have an issue yet.)

![image](https://user-images.githubusercontent.com/11226469/149748749-c4e373a2-730b-44fa-9b20-a4ee6d4f87d0.png)

h) Now for the factor variable I just typed the first value for each category.  But if you look carefully at cell 13 I am still in the cell.
![image](https://user-images.githubusercontent.com/11226469/149749401-55cfb973-bf23-4be5-89f8-cbbaaa3ea91e.png)

i) I then went to Prepare > Replace Values and filled in the NA from above.  It is useful, but it didn't work, because I hadn't left the cell.    I hope this can be fixed and will make an issue.  When you click anywhere else it should close that operation.
j) But anyway, it wouldn't be a problem if we had an undo.  This is issue 4185 and I suggest it should be very simple to add???
j) That worked, because there were 12 values and then another 12.  How would you input the factor variable if there were alternate values?  That's a number of steps.
k) That's also clarified my ideas on what I would like for simple videos.  They are based on tasks, rather than dialogues.  Those tasks may involve a single dialogue, but usually it will be a combination.  So this data entry could be our first of these videos.  If that is how many users start, then we could start here too.
In the video I would like to include the idea of "being lazy" - so the computer does the work!  And I would also include the idea that R-Instat isn't (yet) as easy as some other software for this data entry.  But that's partly because we are happy to co-exist.
The second video can be "Getting data into R-Instat - and out again."
Perhaps there could then be one on "What happens in R-Instat if your data's quite large?  We illustrate with a variety of datasets.  They are all either in the R-Instat library or freely accessible.  (diamonds, rockart, MICS Punjab, IPUMS Kanya - 3 million rows, need a climatic one - including satellite, or reanalysis?)

**Comment by Patowhiz** [comment]
@rdstern I like Bob and your interesting thoughts. 

I think we could implement a very smooth data entry system once we have undo implemented. Without undo, we will never have a proper data entry system in R-Instat. 
One of our biggest strengths is how we do a lot of analytical computations in R, unfortunately that is our downside when it comes to data entry. After undo, our next task could be how to decouple data entry from real time connection to R. We may also simplify this by having a dedicated form for data entry, similar to what we did for climatic data entry. 

Looking forward to having those items solved bit by bit. 

Thanks.
